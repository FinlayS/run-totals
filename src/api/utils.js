export const getUrl = (url) => {
  if (process.env.NODE_ENV !== 'production') {
    return `http://localhost:3001${url}`
  }
  return url
}