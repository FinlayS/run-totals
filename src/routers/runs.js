const express = require('express')
const router = new express.Router()
const auth = require('../middleware/authentication')
const Run = require('../models/run')

router.post('/runs', auth, async (req, res) => {
  const run = new Run({
    ...req.body,
    owner: req.user._id
  })

  try {
    await run.save()
    res.status(201).send(run)

  } catch (e) {
    res.status(400).send(e)
  }
})

router.get('/runs', auth, async (req, res) => {
  try {
    const runs = await Run.find({ owner: req.user._id })
    res.send(runs)

  } catch (e) {
    res.status(500).send()
  }
})

router.get('/runs/:id', auth, async  (req, res) => {
  const _id = req.params.id

  try {
    const run = await Run.findOne({ _id, owner: req.user._id })

    if (!run) {
      return res.status(404).send()
    }

    res.send(run)
  } catch (e) {
    res.status(500).send()
  }
})

router.patch('/runs/:id', auth, async (req, res) => {
  const _id = req.params.id
  const updates = Object.keys(req.body)
  const allowedUpdates = ["description", , "date"]
  const isValidOperation = updates.every((update ) => allowedUpdates.includes(update))

  if (!isValidOperation) {
    return res.status(400).send({ error: "invalid update" })
  }

  try {
    const run = await Run.findOne({ _id, owner: req.user._id })

    if (!run) {
      return res.status(404).send()
    }

    updates.forEach((update) => run[update] = req.body[update])
    await run.save()

    res.send(run)
  } catch(e) {
    return res.status(400).send(e)
  }
})

router.delete('/runs/:id', auth, async (req, res) => {
  const _id = req.params.id

  try {
    const run = await Run.findOneAndDelete({ _id, owner: req.user._id })

    if (!run) {
      return res.status(404).send()
    }

    res.send(run)
  } catch(e) {
    res.status(500).send()
  }
})

module.exports = router
