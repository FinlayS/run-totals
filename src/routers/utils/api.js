import axios from 'axios'

const getUrl = (url) => {
  if (process.env.NODE_ENV !== 'production') {
    const newUrl = `http://localhost:3001${url}`
    return newUrl
  }
  return url
}

export const userLogout = async () => {
  try {
    await axios.post(getUrl('/users/logout'),
      {},
      {
        headers: {
          "Authorization": `Bearer ${localStorage.getItem("token")}`
        }
      }
    )
    localStorage.removeItem("token")

  } catch (e) {
    console.log(e.data)
  }
}
