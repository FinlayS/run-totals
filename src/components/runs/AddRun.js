import React, {useContext, useState} from 'react';
import {Button, Modal, Row, Col, Container} from 'react-bootstrap';
import NumberFormat from "react-number-format";
import moment from 'moment';

import {postRun} from '../../routers/api/runs'
import RunContext from '../../context/runContext';

const AddRun = () => {
  const today = moment().format('DD/MM/YY')
  const {dispatch} = useContext(RunContext)
  const [description, setDescription] = useState('')
  const [date, setDate] = useState(today)
  const [id, setId] = useState('')
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const addRun = async (e) => {

    const runDate = moment(date, 'DD/MM/YY').valueOf()

    let run
    e.preventDefault()

    try {
      run = await postRun({description, runDate})
    } catch (e) {
      console.log(e.data)
    }
    if (run) {
      const _id = run._id
      dispatch(
        {
          type: 'ADD_RUN',
          description,
          runDate,
          _id
        })
      setDescription('')
      setId('')
      setDate(today)
      setShow(false)
    }
  }

  return (
    <>
      <Button variant='primary' onClick={handleShow}>
        Add run
      </Button>
      <Modal show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Add a new run</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <Row className='form-row'>
            <Col className='col-md-8 mb-3 md-form'>
              <label htmlFor='runDescription'>Describe your run</label>
              <input
                type='text'
                className='input-group'
                id='runDescription'
                value={description}
                onChange={(e) => setDescription(e.target.value)}
              />
            </Col>

            <Col className='col-md-4 mb-1 md-form'>
              <label htmlFor='runDate'>Date</label>
              <NumberFormat
                format="##/##/##"
                placeholder="DD/MM/YY"
                mask={['D', 'D', 'M', 'M', 'Y', 'Y']}
                className='input-group'
                id='runDate'
                value={date}
                onChange={(e) => setDate(e.target.value)}
              />
            </Col>
          </Row>
        </Modal.Body>

        <Modal.Footer>
          <Button variant='secondary' onClick={handleClose}>
            Close
          </Button>
          <Button variant='primary' onClick={addRun} type='submit'>
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>

    </>
  )
}

export default AddRun
