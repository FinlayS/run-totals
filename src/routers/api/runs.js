import axios from 'axios';
import { getUrl } from './utils';

export const getRuns = async () => {
  let response
  try {
    response = await axios.get(getUrl('/runs'),
      {
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      }
    )
  } catch (e) {
    console.log(e.data)
  }
  return response.data
}

export const postRun = async (data) => {
  let response
  try {
    response = await axios.post(getUrl('/runs'),
      {
        description: data.description,
        date: data.date,
        runDate: data.runDate
      },
      {
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      }
    )
  } catch (e) {
    console.log(e.data)
  }
  return response.data
}

export const deleteRun = async (id) => {
  let response
  try {
    response = await axios.delete(getUrl(`/runs/${id}`),
      {
        headers: {
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      }
    )
  } catch (e) {
    console.log(e.data)
  }
  return response.data
}
