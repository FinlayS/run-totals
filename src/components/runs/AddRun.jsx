import React, { useContext, useState } from 'react';
import { Button, Modal, Row, Col } from 'react-bootstrap';
import NumberFormat from "react-number-format";
import moment from 'moment';

import { postRun } from '../../api/runs'
import RunContext from '../../context/runContext';
import {CloseIcon, SaveIcon} from "../../../public/icons/icons";

const AddRun = () => {
  const today = moment().format('DD/MM/YY')
  const {dispatchRuns} = useContext(RunContext)
  const [description, setDescription] = useState('')
  const [date, setDate] = useState(today)
  const [, setId] = useState('')
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const addRun = async (e) => {

    const runDate = moment(date, 'DD/MM/YY').valueOf()

    let run
    e.preventDefault()

    try {
      run = await postRun({description, runDate})
    } catch (e) {
      console.log(e.data)
    }
    if (run) {
      const _id = run._id
      dispatchRuns(
        {
          type: 'ADD_RUN',
          description,
          runDate,
          _id
        })
      setDescription('')
      setId('')
      setDate(today)
      setShow(false)
    }
  }

  return (
    <>
      <Button variant='primary' onClick={handleShow}>
        Add run
      </Button>
      <Modal show={show} onHide={handleClose}>
        <Modal.Header>
          <Modal.Title>Add a new run</Modal.Title>
          <Button
            variant='secondary'
            size={"sm"}
            onClick={handleClose}
          >
            <CloseIcon/>
          </Button>
        </Modal.Header>

        <Modal.Body>
          <Row className='form-row'>
            <Col className='col-9'>
              <label htmlFor='runDescription'>Describe your run</label>
              <input
                type='text'
                className='input-group'
                id='runDescription'
                value={description}
                onChange={(e) => setDescription(e.target.value)}
              />
            </Col>

            <Col className='center-self'>
              <label htmlFor='runDate'>Date</label>
              <NumberFormat
                format="##/##/##"
                placeholder="DD/MM/YY"
                mask={['D', 'D', 'M', 'M', 'Y', 'Y']}
                className='input-group center-text'
                id='runDate'
                value={date}
                onChange={(e) => setDate(e.target.value)}
              />
            </Col>
          </Row>
        </Modal.Body>

        <Modal.Footer>
          <Button
            variant='success'
            type='submit'
            onClick={addRun} >
            <SaveIcon/>
            &nbsp; Save
          </Button>
        </Modal.Footer>
      </Modal>

    </>
  )
}

export default AddRun
