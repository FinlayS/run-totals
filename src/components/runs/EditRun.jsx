import React, { useContext, useState } from 'react';
import { Button, Modal, Row, Col } from 'react-bootstrap';
import NumberFormat from "react-number-format";
import moment from 'moment';

import { patchRun } from '../../api/runs'
import RunContext from '../../context/runContext';

const EditRun = ({ run }) => {
  const { dispatchRuns } = useContext(RunContext)
  const [description, setDescription] = useState(run.description)
  const [date, setDate] = useState(moment(run.runDate).format('DD/MM/YY'))
  const [runId] = useState(run._id)
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const editRun = async (e) => {
    const runDate = moment(date, 'DD/MM/YY').valueOf()

    let run
    e.preventDefault()

    try {
      run = await patchRun({ description, runDate }, runId)
    } catch (e) {
      console.log(e.data)
    }
    if (run) {
      const _id = run._id
      dispatchRuns(
        {
          type: 'EDIT_RUN',
          description,
          runDate,
          _id
        })
      dispatchRuns(
        {
          type: 'SORT_RUNS',
        })
      setShow(false)
    }
  }

  return (
    <>
      <Button variant='outline-dark' onClick={handleShow}>
        E
      </Button>
      <Modal show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Edit this run</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <Row className='form-row'>
            <Col className='col-md-8 mb-3 md-form'>
              <label htmlFor='runDescription'>Describe your run</label>
              <input
                type='text'
                className='input-group'
                id='runDescription'
                value={description}
                onChange={(e) => setDescription(e.target.value)}
              />
            </Col>

            <Col className='col-md-4 mb-1 md-form'>
              <label htmlFor='runDate'>Date</label>
              <NumberFormat
                format="##/##/##"
                placeholder="DD/MM/YY"
                mask={['D', 'D', 'M', 'M', 'Y', 'Y']}
                className='input-group'
                id='runDate'
                value={date}
                onChange={(e) => setDate(e.target.value)}
              />
            </Col>
          </Row>
        </Modal.Body>

        <Modal.Footer>
          <Button variant='secondary' onClick={handleClose}>
            Close
          </Button>
          <Button variant='primary' onClick={editRun} type='submit'>
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>

    </>
  )
}

export default EditRun
