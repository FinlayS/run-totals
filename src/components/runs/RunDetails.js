import React, { useContext } from 'react';
import { Row, Col, Container } from 'react-bootstrap';

import { deleteRun } from '../../routers/api/runs';
import RunContext from '../../context/runContext';

import Laps from '../laps/Laps';
import RunSummary from './RunSummary';

const RunDetails = ({run}) => {

  const {dispatch} = useContext(RunContext)

  const removeRun = async () => {
    let r = confirm('Confirm you wish to delete this run');
    if (r === true) {
      let response
      try {
        response = await deleteRun(run._id)
      } catch (e) {
        console.log(e.data)
      }
      if (response) {
        dispatch({type: 'REMOVE_RUN', _id: run._id})
      }
    }
  }

  return (
    <Container>
      <Row className='run-row' key={run._id}>
        <Col> {run.description}</Col>
        <Col xs={4}> {run.date}</Col>
        <Col xs={2}>
          <button
            className='btn btn-link'
            data-toggle='tooltip'
            data-placement='right'
            title='Delete Run'
            onClick={removeRun}
          >x
          </button>
        </Col>
      </Row>
      <Laps >{run._id}</Laps>
      <RunSummary id={run._id}/>
    </Container>
  )
}

export default RunDetails;
