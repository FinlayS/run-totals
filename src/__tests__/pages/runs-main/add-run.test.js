import React from "react";
import { act, render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event";

import RunsMain from "../../../pages/runs-main";
import { getRuns } from "../../../api/runs";

jest.mock("../../../api/runs", () => ({ getRuns: jest.fn(), deleteRun: jest.fn() }));
jest.mock("../../../api/laps", () => ({ getLaps: jest.fn() }));

let addRunButton, mainPage

const addRunButtonContainer = async () => {
  addRunButton = screen.getByRole("button", { name: "Add run" });
}

describe("Runs Main: Add first run", () => {
  beforeEach(async () => {
    jest.resetAllMocks()
    getRuns.mockResolvedValueOnce([])
    await act(async () => {
      mainPage = render(<RunsMain/>);
    });
    await addRunButtonContainer()
  })

  it("should match snapshot for empty page", () => {
    expect(mainPage).toMatchSnapshot()
  })

  it("should open 'Add run' modal", async () => {
    await act(async () =>
      userEvent.click(addRunButton)
    )
  })

})
