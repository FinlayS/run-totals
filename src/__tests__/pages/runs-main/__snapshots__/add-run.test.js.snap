// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runs Main: Add first run should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <title>
        Run Totals
      </title>
      <link
        href="../../public/favicon.ico"
        rel="icon"
        type="image/png"
      />
      <nav
        class="navbar navbar-expand-lg navbar-dark bg-dark sticky-top"
        data-testid="nav-bar-header"
      >
        <button
          aria-controls="responsive-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <span
          class="navBar-title navbar-brand"
        >
          Run Totals
        </span>
        <div
          class="navbar-collapse collapse"
          id="responsive-navbar-nav"
        >
          <div
            class="navbar-nav"
          >
            <div
              class="dropdown nav-item"
            >
              <a
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle nav-link"
                href="#"
                id="nav-title"
                role="button"
              >
                Account
              </a>
            </div>
          </div>
        </div>
        <button
          class="btn btn-primary"
          type="button"
        >
          Add run
        </button>
      </nav>
      <div
        class="background-grey"
      >
        <div
          class="container"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <title>
      Run Totals
    </title>
    <link
      href="../../public/favicon.ico"
      rel="icon"
      type="image/png"
    />
    <nav
      class="navbar navbar-expand-lg navbar-dark bg-dark sticky-top"
      data-testid="nav-bar-header"
    >
      <button
        aria-controls="responsive-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <span
        class="navBar-title navbar-brand"
      >
        Run Totals
      </span>
      <div
        class="navbar-collapse collapse"
        id="responsive-navbar-nav"
      >
        <div
          class="navbar-nav"
        >
          <div
            class="dropdown nav-item"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              id="nav-title"
              role="button"
            >
              Account
            </a>
          </div>
        </div>
      </div>
      <button
        class="btn btn-primary"
        type="button"
      >
        Add run
      </button>
    </nav>
    <div
      class="background-grey"
    >
      <div
        class="container"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Runs Main: Edit run should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
